applicationinsights-android/src/main/java/com/microsoft/applicationinsights/library/AutoCollection.java

```diff
@@ -0,0 +1,358 @@
+
+    /**
+     * The singleton INSTANCE of this class
+     */
+    private static AutoCollection instance;
+
+    /**
+     * The tag for logging
+     */
+    private static final String TAG = "AutoCollection";
+
+    /**
+     * A flag which determines whether auto page view tracking has been enabled or not.
+     */
+    private static boolean autoPageViewsEnabled;
+
+    /**
+     * A flag which determines whether session management has been enabled or not.
+     */
+    private static boolean autoSessionManagementEnabled;
+    ;
+
+    /**
+     * A flag that deterines whether we want to auto-track events for foregrounding backgrounding
+     */
+    private static boolean autoAppearanceTrackingEnabled;
+
+    private static boolean hasRegisteredComponentCallbacks;
+
+    private static boolean hasRegisteredLifecycleCallbacks;
+
+    /**
+     * Create a new INSTANCE of the autocollection event tracking
+     *
+     * @param config           the session configuration for session tracking
+     * @param telemetryContext the context, which is needed to renew sessions
+     */
+    protected AutoCollection(ISessionConfig config, TelemetryContext telemetryContext) {
+        this.activityCount = new AtomicInteger(0);
+        this.lastBackground = new AtomicLong(this.getTime());
+        this.config = config;
+        this.telemetryContext = telemetryContext;
+    }
+
+    /**
+     * Initialize the INSTANCE of Autocollection event tracking.
+     *
+     * @param telemetryContext the context, which is needed to renew sessions
+     * @param config           the session configuration for session tracking
+     */
+    protected static void initialize(TelemetryContext telemetryContext, ISessionConfig config) {
+        // note: isLoaded must be volatile for the double-checked LOCK to work
+        if (!AutoCollection.isLoaded) {
+            synchronized (AutoCollection.LOCK) {
+                if (!AutoCollection.isLoaded) {
+                    AutoCollection.isLoaded = true;
+                    AutoCollection.hasRegisteredComponentCallbacks = false;
+                    AutoCollection.hasRegisteredLifecycleCallbacks = false;
+                    AutoCollection.instance = new AutoCollection(config, telemetryContext);
+                }
+            }
+        }
+    }
+
+    /**
+     * @return the INSTANCE of autocollection event tracking or null if not yet initialized
+     */
+    protected static AutoCollection getInstance() {
+        if (AutoCollection.instance == null) {
+            InternalLogging.error(TAG, "getInstance was called before initialization");
+        }
+
+        return AutoCollection.instance;
+    }
+
+    /**
+     * Enables lifecycle event tracking for the provided application
+     *
+     * @param application the application object
+     */
+    @TargetApi(Build.VERSION_CODES.ICE_CREAM_SANDWICH)
+    private static void registerActivityLifecycleCallbacks(Application application) {
```

Comment:
This seems to be wrong: We need to register for LifecycleCallbacks to enable both Auto Sessions as well as Auto Page Views. If both has been disabled, we also need to unregister those callbacks again. The way you register for those callbacks seems to be correct. But I can't find the code, where you keep track of the state of AutoPageViews & AutoSessionManagement to determine, if the callbacks are still needed. I'm pretty sure that was already implemented in former versions.